"growslice: cap out of range" panic in gob encoding

While trying the `-initpeercount` flag (set to a wrong number to see the effects) one of the weave 1.0.3 routers crashed.

Here are the logs:

```
weave 2015/09/18 16:19:16.475139 ->[192.81.217.207:6783] attempting connection
weave 2015/09/18 16:19:16.477903 ->[192.81.217.207:6783] error during connection attempt: dial tcp4 192.81.217.207:6783: connection refused
weave 2015/09/18 16:19:16.532717 ->[192.241.130.141:6783] attempting connection
weave 2015/09/18 16:19:16.543436 ->[192.241.130.141:6783] error during connection attempt: dial tcp4 192.241.130.141:6783: connection refused
weave 2015/09/18 16:19:18.896839 ->[192.81.217.207:6783] attempting connection
weave 2015/09/18 16:19:18.899240 ->[192.81.217.207:6783] error during connection attempt: dial tcp4 192.81.217.207:6783: connection refused
weave 2015/09/18 16:19:19.505453 ->[192.241.130.141:6783] attempting connection
weave 2015/09/18 16:19:19.509241 ->[192.241.130.141:6783] error during connection attempt: dial tcp4 192.241.130.141:6783: connection refused
panic: runtime error: growslice: cap out of range [recovered]
	panic: runtime error: growslice: cap out of range

goroutine 44 [running]:
encoding/gob.catchError(0xc209bb4630)
	/usr/local/go/src/encoding/gob/error.go:38 +0x96
encoding/gob.encUint8Array(0xc20917dd40, 0xc209b99e00, 0x89b100, 0xc209bc0120, 0x57)
	/usr/local/go/src/encoding/gob/encode.go:238 +0x17b
encoding/gob.(*Encoder).encodeSingle(0xc209bb45a0, 0xc209bb45d8, 0xc209173200, 0x89b100, 0xc209bc0120, 0x57)
	/usr/local/go/src/encoding/gob/encode.go:294 +0x17b
encoding/gob.(*Encoder).encode(0xc209bb45a0, 0xc209bb45d8, 0x89b100, 0xc209bc0120, 0x57, 0xc208034280)
	/usr/local/go/src/encoding/gob/encode.go:694 +0x20f
encoding/gob.(*Encoder).EncodeValue(0xc209bb45a0, 0x89b100, 0xc209bc0120, 0x57, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encoder.go:241 +0x613
encoding/gob.(*Encoder).Encode(0xc209bb45a0, 0x89b100, 0xc209bc0120, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encoder.go:168 +0x7f
github.com/weaveworks/weave/router.GobEncode(0xc2092e5e88, 0x3, 0x3, 0x0, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/weave/router/utils.go:80 +0x130
github.com/weaveworks/weave/router.func·016(0x7ff7ac9bc260, 0xc209b81e60)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip_channel.go:157 +0x290
github.com/weaveworks/weave/router.(*GossipSender).run(0xc20943b060)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:59 +0xf6
created by github.com/weaveworks/weave/router.(*GossipSender).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:49 +0xac

goroutine 1 [chan receive, 12 minutes]:
github.com/weaveworks/weave/common.SignalHandlerLoop(0xc2094e7e68, 0x1, 0x1)
	/home/vagrant/src/github.com/weaveworks/weave/common/signals.go:21 +0x2a4
main.main()
	/home/vagrant/src/github.com/weaveworks/weave/releases/v1.0.3/prog/weaver/main.go:170 +0x1d98

goroutine 5 [syscall, 12 minutes]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 6 [select]:
github.com/weaveworks/weave/ipam.(*Allocator).actorLoop(0xc2091223c0, 0xc20800c540)
	/home/vagrant/src/github.com/weaveworks/weave/ipam/allocator.go:461 +0x14a
created by github.com/weaveworks/weave/ipam.(*Allocator).Start
	/home/vagrant/src/github.com/weaveworks/weave/ipam/allocator.go:79 +0x79

goroutine 7 [select]:
github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents(0xc20911ca80, 0xc209108980)
	/home/vagrant/src/github.com/fsouza/go-dockerclient/event.go:176 +0x419
created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).enableEventMonitoring
	/home/vagrant/src/github.com/fsouza/go-dockerclient/event.go:143 +0x16f

goroutine 9 [chan receive]:
github.com/weaveworks/weave/common/docker.func·001()
	/home/vagrant/src/github.com/weaveworks/weave/common/docker/client.go:52 +0x5c
created by github.com/weaveworks/weave/common/docker.(*Client).AddObserver
	/home/vagrant/src/github.com/weaveworks/weave/common/docker/client.go:59 +0x251

goroutine 17 [syscall, 12 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 10 [select]:
github.com/weaveworks/weave/router.(*LocalPeer).actorLoop(0xc2090d5b30, 0xc20800c720)
	/home/vagrant/src/github.com/weaveworks/weave/router/local_peer.go:173 +0x101
created by github.com/weaveworks/weave/router.(*LocalPeer).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/local_peer.go:27 +0x7e

goroutine 11 [select]:
github.com/weaveworks/weave/router.(*Routes).run(0xc20800c4e0, 0xc20800c780, 0xc20800c7e0)
	/home/vagrant/src/github.com/weaveworks/weave/router/routes.go:154 +0x132
created by github.com/weaveworks/weave/router.(*Routes).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/routes.go:44 +0xd4

goroutine 12 [select]:
github.com/weaveworks/weave/router.(*ConnectionMaker).queryLoop(0xc20912c4c0, 0xc20800c840)
	/home/vagrant/src/github.com/weaveworks/weave/router/connection_maker.go:166 +0x162
created by github.com/weaveworks/weave/router.(*ConnectionMaker).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/connection_maker.go:52 +0x7e

goroutine 13 [syscall]:
syscall.Syscall6(0x2d, 0x8, 0xc2091b1e91, 0xffff, 0x0, 0xc2091b1c50, 0xc2091b1c4c, 0x259e325d3b1e1243, 0x564744aecc4faa84, 0x5d441877ed782eef)
	/usr/local/go/src/syscall/asm_linux_amd64.s:46 +0x5
syscall.recvfrom(0x8, 0xc2091b1e91, 0xffff, 0xffff, 0x0, 0xc2091b1c50, 0xc2091b1c4c, 0x58, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:1906 +0x9b
syscall.Recvfrom(0x8, 0xc2091b1e91, 0xffff, 0xffff, 0x0, 0x6, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/syscall/syscall_unix.go:220 +0xbf
net.(*netFD).readFrom(0xc20911cd20, 0xc2091b1e91, 0xffff, 0xffff, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/fd_unix.go:265 +0x38c
net.(*UDPConn).ReadFromUDP(0xc208036110, 0xc2091b1e91, 0xffff, 0xffff, 0xc2093e43f0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/udpsock_posix.go:67 +0x124
github.com/weaveworks/weave/router.(*Router).udpReader(0xc2080c8420, 0xc208036110, 0x7ff7ac9bb088, 0xc2080360f8)
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:255 +0xfb
created by github.com/weaveworks/weave/router.(*Router).listenUDP
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:246 +0x324

goroutine 14 [IO wait, 12 minutes]:
net.(*pollDesc).Wait(0xc20911cdf0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20911cdf0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc20911cd90, 0x0, 0x7ff7ae3fb730, 0xc2091701e8)
	/usr/local/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208036130, 0xc209140038, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:234 +0x4e
github.com/weaveworks/weave/router.func·029()
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:212 +0x5e
created by github.com/weaveworks/weave/router.(*Router).listenTCP
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:220 +0x2e4

goroutine 15 [syscall, locked to thread]:
github.com/google/gopacket/pcap._Cfunc_pcap_next_ex(0x30d03d0, 0xc20912c768, 0xc20912c770, 0x0)
	/home/vagrant/src/github.com/google/gopacket/pcap/:172 +0x43
github.com/google/gopacket/pcap.(*Handle).getNextBufPtrLocked(0xc20912c740, 0xc209155ef8, 0x0, 0x0)
	/home/vagrant/src/github.com/google/gopacket/pcap/pcap.go:306 +0x92
github.com/google/gopacket/pcap.(*Handle).ZeroCopyReadPacketData(0xc20912c740, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/vagrant/src/github.com/google/gopacket/pcap/pcap.go:338 +0xb0
github.com/weaveworks/weave/router.(*PcapIO).ReadPacket(0xc2080360f0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/weave/router/pcap.go:76 +0x67
github.com/weaveworks/weave/router.func·028()
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:144 +0x5d
created by github.com/weaveworks/weave/router.(*Router).sniff
	/home/vagrant/src/github.com/weaveworks/weave/router/router.go:149 +0x40f

goroutine 16 [IO wait]:
net.(*pollDesc).Wait(0xc2092d32c0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2092d32c0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc2092d3260, 0x0, 0x7ff7ae3fb730, 0xc208d79208)
	/usr/local/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208036350, 0xc2092e8cf8, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc208036350, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:244 +0x4c
net/http.(*Server).Serve(0xc20800db00, 0x7ff7ac9bb740, 0xc208036350, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1728 +0x92
net/http.Serve(0x7ff7ac9bb740, 0xc208036350, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1606 +0xa1
main.handleHTTP(0xc2080c8420, 0xc2090d7cb8, 0x5, 0xc2091223c0, 0xa800000, 0xa, 0xc2080360c0)
	/home/vagrant/src/github.com/weaveworks/weave/releases/v1.0.3/prog/weaver/main.go:304 +0x8ff
created by main.main
	/home/vagrant/src/github.com/weaveworks/weave/releases/v1.0.3/prog/weaver/main.go:167 +0x1c88

goroutine 20 [select]:
github.com/weaveworks/weave/router.(*GossipSender).run(0xc209172e60)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:54 +0x22c
created by github.com/weaveworks/weave/router.(*GossipSender).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:49 +0xac

goroutine 19 [select]:
github.com/weaveworks/weave/router.(*LocalConnection).actorLoop(0xc2092e0000, 0xc209178180, 0x0, 0x0)
	/home/vagrant/src/github.com/weaveworks/weave/router/connection.go:375 +0x2d9
github.com/weaveworks/weave/router.(*LocalConnection).run(0xc2092e0000, 0xc209178180, 0xc2091781e0, 0x9d8b01)
	/home/vagrant/src/github.com/weaveworks/weave/router/connection.go:310 +0x382
created by github.com/weaveworks/weave/router.(*LocalConnection).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/connection.go:121 +0xdd

goroutine 21 [select]:
github.com/weaveworks/weave/router.(*GossipSender).run(0xc209172ea0)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:54 +0x22c
created by github.com/weaveworks/weave/router.(*GossipSender).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:49 +0xac

goroutine 22 [select, 10 minutes]:
github.com/weaveworks/weave/router.(*GossipSender).run(0xc209172ee0)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:54 +0x22c
created by github.com/weaveworks/weave/router.(*GossipSender).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:49 +0xac

goroutine 23 [IO wait]:
net.(*pollDesc).Wait(0xc209174220, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc209174220, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc2091741c0, 0xc2092ea000, 0x1000, 0x1000, 0x0, 0x7ff7ae3fb730, 0xc209059d08)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc209140038, 0xc2092ea000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
bufio.(*Reader).fill(0xc2091782a0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Read(0xc2091782a0, 0xc209170240, 0x1, 0x9, 0xc209da6a50, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:174 +0x26c
io.ReadAtLeast(0x7ff7ac9b99f0, 0xc2091782a0, 0xc209170240, 0x1, 0x9, 0x1, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:298 +0xf1
io.ReadFull(0x7ff7ac9b99f0, 0xc2091782a0, 0xc209170240, 0x1, 0x9, 0x20, 0x0, 0x0)
	/usr/local/go/src/io/io.go:316 +0x6d
encoding/gob.decodeUintReader(0x7ff7ac9b99f0, 0xc2091782a0, 0xc209170240, 0x9, 0x9, 0x0, 0x1, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/decode.go:121 +0x99
encoding/gob.(*Decoder).recvMessage(0xc209176200, 0xc)
	/usr/local/go/src/encoding/gob/decoder.go:76 +0x55
encoding/gob.(*Decoder).decodeTypeSequence(0xc209176200, 0xc2090afa00, 0x16)
	/usr/local/go/src/encoding/gob/decoder.go:140 +0x47
encoding/gob.(*Decoder).DecodeValue(0xc209176200, 0x87e8a0, 0xc2090afac0, 0x16, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/decoder.go:208 +0x192
encoding/gob.(*Decoder).Decode(0xc209176200, 0x87e8a0, 0xc2090afac0, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/decoder.go:185 +0x297
github.com/weaveworks/weave/router.(*LocalConnection).receiveTCP(0xc2092e0000, 0xc209176200)
	/home/vagrant/src/github.com/weaveworks/weave/router/connection.go:444 +0x10a
created by github.com/weaveworks/weave/router.(*LocalConnection).run
	/home/vagrant/src/github.com/weaveworks/weave/router/connection.go:309 +0x362

goroutine 29 [chan receive]:
github.com/weaveworks/weave/router.(*Forwarder).run(0xc209186b40, 0xc2093481e0, 0xc209348240)
	/home/vagrant/src/github.com/weaveworks/weave/router/forwarder.go:228 +0x89
created by github.com/weaveworks/weave/router.(*Forwarder).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/forwarder.go:211 +0xce

goroutine 35 [chan receive]:
github.com/weaveworks/weave/router.(*Forwarder).run(0xc2091875e0, 0xc209349920, 0xc209349980)
	/home/vagrant/src/github.com/weaveworks/weave/router/forwarder.go:228 +0x89
created by github.com/weaveworks/weave/router.(*Forwarder).Start
	/home/vagrant/src/github.com/weaveworks/weave/router/forwarder.go:211 +0xce

goroutine 27 [select, 10 minutes]:
github.com/weaveworks/weave/router.(*GossipSender).run(0xc209343c40)
	/home/vagrant/src/github.com/weaveworks/weave/router/gossip.go:54 +0x22c
created by github.com/weaveworks/weave/router.(*GossipSender).Start
```

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -478,8 +478,8 @@ type ipamGossipData struct {
 	alloc *Allocator
 }
 
-func (d *ipamGossipData) Merge(other mesh.GossipData) {
-	// no-op
+func (d *ipamGossipData) Merge(other mesh.GossipData) mesh.GossipData {
+	return d // no-op
 }
 
 func (d *ipamGossipData) Encode() [][]byte {

--- a/mesh/gossip.go
+++ b/mesh/gossip.go
@@ -6,7 +6,7 @@ import (
 
 type GossipData interface {
 	Encode() [][]byte
-	Merge(GossipData)
+	Merge(GossipData) GossipData
 }
 
 type Gossip interface {
@@ -79,8 +79,7 @@ func (sender *GossipSender) Send(data GossipData) {
 	// NB: this must not be invoked concurrently
 	select {
 	case pending := <-sender.cell:
-		pending.Merge(data)
-		sender.cell <- pending
+		sender.cell <- pending.Merge(data)
 	default:
 		sender.cell <- data
 	}

--- a/mesh/router.go
+++ b/mesh/router.go
@@ -127,17 +127,19 @@ func (router *Router) BroadcastTopologyUpdate(update []*Peer) {
 	for _, p := range update {
 		names[p.Name] = void
 	}
-
-	router.TopologyGossip.GossipBroadcast(&TopologyGossipData{
-		peers:  router.Peers,
-		update: names,
-	})
+	router.TopologyGossip.GossipBroadcast(
+		&TopologyGossipData{peers: router.Peers, update: names})
 }
 
-func (d *TopologyGossipData) Merge(other GossipData) {
-	for name := range other.(*TopologyGossipData).update {
-		d.update[name] = void
+func (d *TopologyGossipData) Merge(other GossipData) GossipData {
+	names := make(PeerNameSet)
+	for name := range d.update {
+		names[name] = void
 	}
+	for name := range other.(*TopologyGossipData).update {
+		names[name] = void
+	}
+	return &TopologyGossipData{peers: d.peers, update: names}
 }
 
 func (d *TopologyGossipData) Encode() [][]byte {

--- a/mesh/surrogate_gossiper.go
+++ b/mesh/surrogate_gossiper.go
@@ -12,8 +12,12 @@ func (d *SurrogateGossipData) Encode() [][]byte {
 	return d.messages
 }
 
-func (d *SurrogateGossipData) Merge(other GossipData) {
-	d.messages = append(d.messages, other.(*SurrogateGossipData).messages...)
+func (d *SurrogateGossipData) Merge(other GossipData) GossipData {
+	o := other.(*SurrogateGossipData)
+	messages := make([][]byte, 0, len(d.messages)+len(o.messages))
+	messages = append(messages, d.messages...)
+	messages = append(messages, o.messages...)
+	return &SurrogateGossipData{messages: messages}
 }
 
 // SurrogateGossiper ignores unicasts and relays broadcasts and gossips.

--- a/nameserver/entry.go
+++ b/nameserver/entry.go
@@ -247,11 +247,7 @@ type GossipData struct {
 
 func (g *GossipData) Merge(o mesh.GossipData) mesh.GossipData {
 	other := o.(*GossipData)
-	gossip := &GossipData{
-		Entries:   make(Entries, len(g.Entries)),
-		Timestamp: g.Timestamp,
-	}
-	copy(gossip.Entries, g.Entries)
+	gossip := g.copy()
 	gossip.Entries.merge(other.Entries)
 	if gossip.Timestamp < other.Timestamp {
 		gossip.Timestamp = other.Timestamp
@@ -270,13 +266,17 @@ func (g *GossipData) Decode(msg []byte) error {
 }
 
 func (g *GossipData) Encode() [][]byte {
-	// Make a copy so we can sort: all outgoing data is sent in case-sensitive order
-	g2 := GossipData{Timestamp: g.Timestamp, Entries: make(Entries, len(g.Entries))}
-	copy(g2.Entries, g.Entries)
+	g2 := g.copy()
 	sort.Sort(CaseSensitive(g2.Entries))
 	buf := &bytes.Buffer{}
 	if err := gob.NewEncoder(buf).Encode(g2); err != nil {
 		panic(err)
 	}
 	return [][]byte{buf.Bytes()}
+}
+
+func (g *GossipData) copy() *GossipData {
+	g2 := &GossipData{Timestamp: g.Timestamp, Entries: make(Entries, len(g.Entries))}
+	copy(g2.Entries, g.Entries)
+	return g2
 }

--- a/nameserver/entry.go
+++ b/nameserver/entry.go
@@ -245,12 +245,18 @@ type GossipData struct {
 	Entries
 }
 
-func (g *GossipData) Merge(o mesh.GossipData) {
+func (g *GossipData) Merge(o mesh.GossipData) mesh.GossipData {
 	other := o.(*GossipData)
-	g.Entries.merge(other.Entries)
-	if g.Timestamp < other.Timestamp {
-		g.Timestamp = other.Timestamp
+	gossip := &GossipData{
+		Entries:   make(Entries, len(g.Entries)),
+		Timestamp: g.Timestamp,
 	}
+	copy(gossip.Entries, g.Entries)
+	gossip.Entries.merge(other.Entries)
+	if gossip.Timestamp < other.Timestamp {
+		gossip.Timestamp = other.Timestamp
+	}
+	return gossip
 }
 
 func (g *GossipData) Decode(msg []byte) error {

--- a/nameserver/entry_test.go
+++ b/nameserver/entry_test.go
@@ -117,7 +117,7 @@ func TestGossipDataMerge(t *testing.T) {
 	g1 := GossipData{Entries: makeEntries("AcDf")}
 	g2 := GossipData{Entries: makeEntries("BEf")}
 
-	g1.Merge(&g2)
+	g3 := g1.Merge(&g2).(*GossipData)
 
-	require.Equal(t, GossipData{Entries: makeEntries("ABcDEf")}, g1)
+	require.Equal(t, GossipData{Entries: makeEntries("ABcDEf")}, *g3)
 }

